src/metabase/query_processor/card.clj

```diff
@@ -34,34 +32,25 @@
 
 (set! *warn-on-reflection* true)
 
-(defn- query-magic-ttl
-  "Compute a 'magic' cache TTL time (in seconds) for `query` by multipling its historic average execution times by the
-  `query-caching-ttl-ratio`. If the TTL is less than a second, this returns `nil` (i.e., the cache should not be
-  utilized.)"
-  [query]
-  (when-let [average-duration (query/average-execution-time-ms (qp.util/query-hash query))]
-    (let [ttl-seconds (Math/round (float (/ (* average-duration (public-settings/query-caching-ttl-ratio))
-                                            1000.0)))]
-      (when-not (zero? ttl-seconds)
-        (log/info (trs "Question''s average execution duration is {0}; using ''magic'' TTL of {1}"
-                       (u/format-milliseconds average-duration) (u/format-seconds ttl-seconds))
-                  (u/emoji "💾"))
-        ttl-seconds))))
-
-(defenterprise granular-ttl
-  "Returns the granular cache ttl (in seconds) for a card. On EE, this first checking whether there is a stored value
-   for the card, dashboard, or database (in that order of decreasing preference). Returns nil on OSS."
-  metabase-enterprise.advanced-config.caching
-  [_card _dashboard _database])
-
-(defn- ttl-hierarchy
-  "Returns the cache ttl (in seconds), by first checking whether there is a stored value for the database,
-    dashboard, or card (in that order of increasing preference), and if all of those don't exist, then the
-    `query-magic-ttl`, which is based on average execution time."
-  [card dashboard database query]
+(defenterprise granular-cache-strategy
+  "Returns cache strategy for a card. On EE, this checks the hierarchy for the card, dashboard, collection, or
+   database (in that order). Returns nil on OSS."
+  metabase-enterprise.caching.strategies
+  [_card _dashboard-id])
+
+(defn- cache-strategy
+  [card dashboard-id]
   (when (public-settings/enable-query-caching)
-    (or (granular-ttl card dashboard database)
-        (query-magic-ttl query))))
+    (or (granular-cache-strategy card dashboard-id)
+        {:type         :ttl
+         :multiplier   (public-settings/query-caching-ttl-ratio)
+         :min_duration (long (* (public-settings/query-caching-min-ttl) 1000))})))
+
+(defn- enrich-strategy [strategy query]
+  (case (:type strategy)
+    :ttl (let [et (query/average-execution-time-ms (qp.util/query-hash query))]
+           (assoc strategy :execution-time (or et 0)))
```

Comment:
Is it not worth renaming this key to indicate the units, and that it is an average? Naively one might expect it to be the last run time.

Reply:
renamed to `avg-execution-ms`